### PR TITLE
[Core] Add PAG support for PixArtSigma 

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Part of #8785. 

I think PixArt Sigma is way more popular than PixArt Alpha (longer sequence length, better quality in general, etc.), it makes sense to just add PAG to PixArt Sigma. If there's a need, I think we could open it up to community. 

## TODO 

- [x] Tests
- [x] Docs

## Code

```python
from diffusers import PixArtSigmaPAGPipeline
import torch 

pipe = PixArtSigmaPAGPipeline.from_pretrained(
    "PixArt-alpha/PixArt-Sigma-XL-2-1024-MS", 
    torch_dtype=torch.float16, 
    pag_applied_layers=[14]
).to("cuda")

image = pipe(
    "A small cactus with a happy face in the Sahara desert.", 
    guidance_scale=0.0, 
    pag_scale=2.0, 
    generator=torch.manual_seed(0)
).images[0]
```

A bit of ablation study is in the comment below. 

@asomoza would appreciate it if you could run some experiments with it :) 

@sunovivid for awareness.

### CODE REVIEW

**Comment by sayakpaul** [comment]
# Ablation

<details>
<summary><h2>Code</h2></summary>

```python
import torch
import argparse
from diffusers import PixArtSigmaPAGPipeline, PixArtSigmaPipeline

def load_pipeline(args):
    if args.pag:
        pipe = PixArtSigmaPAGPipeline.from_pretrained(
            "PixArt-alpha/PixArt-Sigma-XL-2-1024-MS", torch_dtype=torch.float16, pag_applied_layers=args.pag_applied_layers
        ).to("cuda")
    else:
        pipe = PixArtSigmaPipeline.from_pretrained("PixArt-alpha/PixArt-Sigma-XL-2-1024-MS", torch_dtype=torch.float16).to("cuda")
    return pipe 

def run_pipeline(pipe, args):
    if args.pag:
        image = pipe(
            args.prompt, guidance_scale=args.cfg, pag_scale=args.pag_scale, generator=torch.manual_seed(0)
        ).images[0]
    else:
        image = pipe(args.prompt, generator=torch.manual_seed(0)).images[0]
    
    img_name = "_".join(args.prompt.split(" "))
    if args.pag:
        img_name += "_pag"
        pag_applied_layers = list(map(str, args.pag_applied_layers))
        pag_applied_layers = "_".join(pag_applied_layers)
        img_name += f"_cfg@{args.cfg}_pg@{args.pag_scale}_layers@{pag_applied_layers}"
    img_name += ".png"
    image.save(img_name)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--pag", type=int, default=0)
    parser.add_argument("--prompt", type=str, default="A small cactus with a happy face in the Sahara desert.")
    parser.add_argument("--cfg", type=float, default=0.0)
    parser.add_argument("--pag_scale", type=float, default=4.0)
    parser.add_argument("--pag_applied_layers", metavar="N", type=int, nargs="*", default=[14])
    args = parser.parse_args()

    pipe = load_pipeline(args)
    run_pipeline(pipe, args)
```
</details>

## Results

Prompt: _A small cactus with a happy face in the Sahara desert_

|W/O PAG|PAG|
|---|---|
|||

Prompt: _Astronaut on Mars During sunset_

|W/O PAG|PAG_CFG_0_PG_2_LAYERS_14|PAG_CFG_0_PG_2_LAYERS_13_14|PAG_CFG_2_PG_2_LAYERS_14|
|---|---|---|---|
||||

File: src/diffusers/pipelines/pag/pag_utils.py [hunk 1]
```diff
@@ -258,3 +258,91 @@ def pag_attn_processors(self):
             if proc.__class__ in (PAGCFGIdentitySelfAttnProcessor2_0, PAGIdentitySelfAttnProcessor2_0):
                 processors[name] = proc
         return processors
+
+
+class PixArtPAGMixin(PAGMixin):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pag/pag_utils.py:263]: Let me know if I should copy-paste the other methods from `PAGMixin` and not make it a subclass of `PAGMixin`. 

`PixArtPAGMixin` differs only in terms of the methods I implemented here.

Comment (sunovivid) [reply src/diffusers/pipelines/pag/pag_utils.py:263]: I'm also curious about this. subclass looks neater than copy-paste.

Comment (yiyixuxu) [reply src/diffusers/pipelines/pag/pag_utils.py:263]: I would prefer not to have a Mixin nested into another Mixin if it's possible 
we can either 
1. copy-paste the common methods from `PAGMixin` (maybe we can rename it to `SDPAGMixin`)
2. we just keep on common `PAGMixin` and move these methods that're potentially model-specific to pipeline methods

**Comment by yiyixuxu** [comment]
maybe we can try the robot prepare meal example? https://huggingface.co/docs/diffusers/main/en/using-diffusers/pag#general-tasks

I cannot tell if there is an improvement or not in the example provided (but I'm also not very good at looking at this, so cc @asomoza for help here too)

File: src/diffusers/pipelines/pag/pag_utils.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pag_utils.py:263]: I will go with the first one. I will defer any refactoring to @yiyixuxu if she feels the need to do one.

**Comment by sayakpaul** [comment]
Yeah I am not good at that either. So, requested @asomoza to chime in :D 

Here are the insect cooking meal results. 

| No PAG | CFG_1_PG_3_L14 | CFG_4.5_PG_3_L13_14_15 | CFG_4.5_PG_3_L14 |
|----------|----------|----------|----------|
| Image 1 | Image 2 | Image 3 | Image 4 |

`CFG_1_PG_3_L14` means we are applying PAG with the following config:

* CFG of 1.0
* PAG scale of 3.0
* `pag_applied_layers` is set to [14]

File: src/diffusers/pipelines/pag/pag_utils.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pag_utils.py]: Done in [27fceb3](https://github.com/huggingface/diffusers/pull/8921/commits/27fceb3b2608835e91b0e417f0560e39d259a7c5). LMK.

**Comment by asomoza** [comment]
This one is the most interesting one of the new models, it definitely helps with the generation, it cleans the image and fixes the errors, but as with kolors, it takes away some details, so it depends on what you want.

For example in this image, it fixes the hands of the robot and the shape of the bowls:

|W/O PAG|PAG|
|---|---|
|||

Also I found the time to play with the model and it's a really good model, I'm impressed with the generations it can do.

Anyway, what I found interesting is how the PAG layers behave, they affect a lot of the generation, for example:

|L1|L2|L6|
|---|---|---|
|20240722232322_1321339354_layer_1|20240722232333_1321339354_layer_2|20240722232418_1321339354_layer_6|
|L18|L19|L21|
|20240722232633_1321339354_layer_18|20240722232644_1321339354_layer_19|20240722232707_1321339354_layer_21|

As with SDXL, I can see some relation of the layers with the generations but I'll need to play more with it to be sure. For example the 19 and 21 layers seems to affect more the background and some others seems to affect more the shape or the colors. Also the 1 and 6 layers can be used for the composition with a second pass over them.

**Comment by sayakpaul** [comment]
Thank you very much, Alvaro! So, what I am hearing is that it makes sense to have PAG supported for PixArt Sigma, yeah? 

Cc @lawrence-cj too you might find it nice.

File: src/diffusers/pipelines/pag/pag_utils.py [hunk 2]
```diff
@@ -258,3 +258,185 @@ def pag_attn_processors(self):
...
+
+        # If it is not a valid layer index, raise a ValueError
+        raise ValueError(f"Pag layer should only contain block index. Accept number string like '3', got {layer}.")
+
+    def _set_pag_attn_processor(self, pag_applied_layers, do_classifier_free_guidance):
+        r"""
+        Set the attention processor for the PAG layers.
+        """
+        if do_classifier_free_guidance:
+            pag_attn_proc = PAGCFGIdentitySelfAttnProcessor2_0()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:286]: I think if we make `pag_attn_proc` a attribute of PAG pipeline, `PixArtPAGMixin` can be used by all transformer based model, no? sd3, hunyuan? we can rename it to `TransformerPAGMixin` or so

File: src/diffusers/pipelines/pag/pag_utils.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/pag/pag_utils.py]: cc @a-r-r-o-w here too, I think we can reuse this one for hunyuan too

**Comment by yiyixuxu** [comment]
@asomoza very nice! thanks!

File: src/diffusers/pipelines/pag/pag_utils.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pag_utils.py:286]: Maybe we will know better once we have different variants as IMO that gives a better view of things that could be refactored. 

I would prefer to do that in a separate PR.

Comment (yiyixuxu) [reply src/diffusers/pipelines/pag/pag_utils.py:286]: ok with the plan! 
no need to open a separate PR 
@a-r-r-o-w can refactor in hunyun PR!

Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pag_utils.py:286]: yes, sorry if ive missed something! 😅 i had some ideas about reimplementing PAG, so please don't be blocked on any merges by me. also need to analyse a different setting in hunyuan which I'll do over the weekend...

Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pag_utils.py:286]: Cool!

**Comment by sayakpaul** [comment]
@yiyixuxu can I add test and docs and get it ready for final review?

**Comment by yiyixuxu** [comment]
@sayakpaul sure ! looks good to me
let's merge this soon

File: tests/pipelines/pag/test_pag_pixart_sigma.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+
+enable_full_determinism()
+
+
+class PixArtSigmaPAGPipelineFastTests(PipelineTesterMixin, unittest.TestCase):
+    pipeline_class = PixArtSigmaPAGPipeline
+    params = TEXT_TO_IMAGE_PARAMS.union({"pag_scale", "pag_adaptive_scale"})
+    params = set(params)
+    params.remove("cross_attention_kwargs")
```
Comment (sayakpaul) [inline tests/pipelines/pag/test_pag_pixart_sigma.py:55]: LMK if this needs to be handled differently. PixArt doesn't have `cross_attention_kwargs` and doesn't need to have yet.

File: tests/pipelines/pag/test_pag_pixart_sigma.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        }
+
+        output_loaded = pipe_loaded(**inputs)[0]
+
+        max_diff = np.abs(to_np(output) - to_np(output_loaded)).max()
+        self.assertLess(max_diff, 1e-4)
+
+    # Because the PAG PixArt Sigma has `pag_applied_layers`.
+    # Also, we shouldn't be doing `set_default_attn_processor()` after loading
+    # the pipeline with `pag_applied_layers`.
```
Comment (sayakpaul) [inline tests/pipelines/pag/test_pag_pixart_sigma.py:267]: Let me know if this and subsequent tests should be handled differently.

**Comment by sayakpaul** [comment]
@yiyixuxu I have left some questions for you regarding the tests. LMK.

File: tests/pipelines/pag/test_pag_pixart_sigma.py (hunk 3 above)
Comment (yiyixuxu) [reply tests/pipelines/pag/test_pag_pixart_sigma.py:55]: ok!

File: tests/pipelines/pag/test_pag_pixart_sigma.py (hunk 4 above)
Comment (yiyixuxu) [reply tests/pipelines/pag/test_pag_pixart_sigma.py:267]: looks good!

**Comment by yiyixuxu** [comment]
@sayakpaul looks good to me! feel free to merge once the conflicts are resolved

**Pull request merged**